chain/builder.go

```diff
@@ -297,23 +305,193 @@ func BuildBlock(
 			tsv.Commit()
 			b.Txs = append(b.Txs, anchorTxs...)
+			// Add txs from anchor to the set, so we can skip them during local build process.
+			for _, tx := range b.Txs {
+				includedTxsSet.Add(tx.ID())
+			}
+			bwConsumerd = feeManager.LastConsumed(fees.Bandwidth)
 			results = append(results, anchorTxResults...)
+			isAnchorBuildSuccessful = true
 		}
 	}
-
 skipAnchor:
+	if vm.IsArcadiaConfigured() {
+		lbwUnits := maxUnits[fees.Bandwidth] - bwConsumerd - 50*units.KiB
+		txs, err := GetArcadiaTxs(ctx, vm, r, lbwUnits, nextHght)
+		if err != nil {
+			goto skipArcadia
+		}
+
+		ctx, executeSpan := vm.Tracer().Start(ctx, "chain.BuildBlock.ExecuteArcadia")
+		ntxs := len(txs)
+		vm.Logger().Debug("arcadia batch info", zap.Int("size", ntxs))
+
+		arcadiaStateKeys := make(state.Keys)
+		for _, tx := range txs {
+			statekeys, err := tx.StateKeys(sm)
+			if err != nil {
+				vm.Logger().Error("unable to get statekeys of tx", zap.Error(err))
+				goto skipArcadia
+			}
+			for k, perm := range statekeys {
+				arcadiaStateKeys.Add(k, perm)
+			}
+		}
+		var (
+			storage  = make(map[string][]byte, len(arcadiaStateKeys))
+			toLookup = make([]string, 0, len(arcadiaStateKeys))
+		)
+
+		cacheLock.RLock()
+		for k := range arcadiaStateKeys {
+			if v, ok := cache[k]; ok {
+				if v.exists {
+					storage[k] = v.v
+				}
+				continue
+			}
+			toLookup = append(toLookup, k)
+		}
+		cacheLock.RUnlock()
+
+		// Fetch keys from disk
+		var toCache map[string]*fetchData
+		if len(toLookup) > 0 {
+			toCache = make(map[string]*fetchData, len(toLookup))
+			for _, k := range toLookup {
+				v, err := parentView.GetValue(ctx, []byte(k))
+				if errors.Is(err, database.ErrNotFound) {
+					toCache[k] = &fetchData{nil, false, 0}
+					continue
+				} else if err != nil {
+					vm.Logger().Error("unable to get key", zap.String("key", k), zap.Error(err))
+					goto skipArcadia
+				}
+				// We verify that the [NumChunks] is already less than the number
+				// added on the write path, so we don't need to do so again here.
+				numChunks, ok := keys.NumChunks(v)
+				if !ok {
+					vm.Logger().Error("unable to calculate num of chunks for k", zap.String("key", k))
+					goto skipArcadia
+				}
+				toCache[k] = &fetchData{v, true, numChunks}
+				storage[k] = v
+			}
+
+			// Update key cache regardless of whether exit is graceful
+			cacheLock.Lock()
+			for k := range toCache {
+				cache[k] = toCache[k]
+			}
+			cacheLock.Unlock()
+		}
+
+		var arcadiaL sync.Mutex
+		arcadiaTxs := make([]*Transaction, 0, len(txs))
+		arcadiaTxResults := make([]*Result, 0, len(results))
+		e := executor.New(ntxs, vm.GetTransactionExecutionCores(), MaxKeyDependencies, vm.GetExecutorBuildRecorder())
+		// start a new view here, only commit when all txs succeed
+		tsv := ts.NewView(arcadiaStateKeys, storage)
+		for _, ltx := range txs {
+			txsAttempted++
+			tx := ltx
+			// TODO: parallem this, currently `TState` can't be rolled back while `TState_View` can
```

Comment:
parallelize

Reply:
redundant comment carried over. It's actually parallelized.